### PR TITLE
Avoid duplication of mod metadata in memory

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
I'd like mod storage to act more like a proper key-value database. It shouldn't store all the data in memory. With this PR, using the SQLite3 backend, the data is no longer kept in memory (unless SQLite3 chooses to do so.) The metadata interface has been split into `IMetadata` for the basic interface and `SimpleMetadata` for in-memory metadata. `ModMetadata` extends `IMetadata`. The interface only requires all data to be in memory when comparing two metadata maps and when converting metadata to a table. I could avoid this, but I'd rather not implement custom iterators and all that BS.

With this PR, mod writers can store huge amounts of data in mod storage without worrying about RAM limits, provided they are using the SQLite3 backend.

## To do

This PR is Ready for Review.

- [x] C++ unit tests
- [x] Lua unit tests
- [x] `minetest.features` entry

## How to test

There are unit tests in Lua and C++. You can also test with mods that use mod storage and other metadata.

### CODE REVIEW

**Comment by Zughy** [comment]
@TurkeyMcMac rebase needed

File: doc/lua_api.txt [hunk 1]
```diff
@@ -4868,6 +4868,9 @@ Utilities
           get_sky_as_table = true,
           -- VoxelManip:get_light_data accepts an optional buffer argument (5.7.0)
           get_light_data_buffer = true,
+          -- When using an appropriate backend (not the deprecated "files" backend),
+          -- mod storage data is not necessarily all kept in RAM (5.7.0)
+          mod_storage_on_disk = true,
```
Comment (sfan5) [inline doc/lua_api.txt:4873]: fuzzy as it is what is the practical application?

File: src/content/mods.cpp [hunk 2]
```diff
@@ -220,26 +220,34 @@ std::vector<ModSpec> flattenMods(const std::map<std::string, ModSpec> &mods)
...
+}
+
+const StringMap &ModMetadata::getStrings(StringMap *place) const
+{
+	place->clear();
+	m_database->getModEntries(m_mod_name, place);
+	return *place;
+}
+
+const std::string *ModMetadata::getStringRaw(const std::string &name, std::string *place) const
```
Comment (sfan5) [inline src/content/mods.cpp:250]: return `const &` for consistency?

File: src/database/database-dummy.cpp [hunk 3]
```diff
@@ -92,6 +92,32 @@ bool Database_Dummy::getModEntries(const std::string &modname, StringMap *storag
...
+bool Database_Dummy::getModEntry(const std::string &modname,
+	const std::string &key, std::string *value)
+{
+	auto mod_pair = m_mod_meta_database.find(modname);
+	if (mod_pair == m_mod_meta_database.end())
+		return false;
+	const StringMap &meta = mod_pair->second;
+
+	auto pair = meta.find(key);
+	if (pair == meta.end()) {
```
Comment (sfan5) [inline src/database/database-dummy.cpp:104]: condition is wrong way around

File: src/database/database-dummy.cpp [hunk 4]
```diff
@@ -112,6 +138,16 @@ bool Database_Dummy::removeModEntry(const std::string &modname, const std::strin
 	return false;
 }
 
+bool Database_Dummy::removeModEntries(const std::string &modname)
+{
+	auto mod_pair = m_mod_meta_database.find(modname);
+	if (mod_pair != m_mod_meta_database.end() && !mod_pair->second.empty()) {
```
Comment (sfan5) [inline src/database/database-dummy.cpp:144]: is returning false when already empty part of the interface contract?

File: src/metadata.cpp [hunk 5]
```diff
@@ -21,95 +21,110 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 {
-	StringMap::const_iterator it = m_stringvars.find(name);
-	if (it == m_stringvars.end()) {
+	const std::string *raw = getStringRaw(name, &str);
+	if (!raw)
 		return false;
-	}
 
-	str = resolveString(it->second, recursion);
+	const std::string &resolved = resolveString(*raw, &str, recursion);
```
Comment (sfan5) [inline src/metadata.cpp:64]: does this work safely if `raw == &str`?

File: src/metadata.h [hunk 6]
```diff
@@ -43,21 +42,76 @@ class Metadata
...
 	bool getStringToRef(const std::string &name, std::string &str, u16 recursion = 0) const;
-	virtual bool setString(const std::string &name, const std::string &var);
+
+	// Returns whether the metadata was (potentially) changed.
+	virtual bool setString(const std::string &name, const std::string &var) = 0;
+
 	inline bool removeString(const std::string &name) { return setString(name, ""); }
-	const StringMap &getStrings() const
+
+	// May put strings in *place and return a reference to these strings.
```
Comment (sfan5) [inline src/metadata.h:60]:
```suggestion
	// May (not must!) put strings in `place` and return a reference to these strings.
```

File: src/metadata.h [hunk 7]
```diff
@@ -43,21 +42,76 @@ class Metadata
 	// Key-value related
 	//
 
-	size_t size() const;
-	bool contains(const std::string &name) const;
-	const std::string &getString(const std::string &name, u16 recursion = 0) const;
+	virtual bool contains(const std::string &name) const = 0;
+
+	// May put a string in *place and return a reference to that string.
```
Comment (sfan5) [inline src/metadata.h:47]:
```suggestion
	// May (not must!) put string in `place` and return a reference to that string.
```

File: src/script/lua_api/l_nodemeta.cpp [hunk 8]
```diff
@@ -145,13 +145,13 @@ void NodeMetaRef::handleToTable(lua_State *L, Metadata *_meta)
...
 // from_table(self, table)
-bool NodeMetaRef::handleFromTable(lua_State *L, int table, Metadata *_meta)
+bool NodeMetaRef::handleFromTable(lua_State *L, int table, IMetadata *_meta)
 {
 	// fields
 	if (!MetaDataRef::handleFromTable(L, table, _meta))
 		return false;
 
-	NodeMetadata *meta = (NodeMetadata*) _meta;
+	NodeMetadata *meta = dynamic_cast<NodeMetadata*>(_meta);
```
Comment (sfan5) [inline src/script/lua_api/l_nodemeta.cpp:154]: should add an assert here

File: src/script/lua_api/l_storage.cpp [hunk 9]
```diff
@@ -28,23 +28,12 @@ int ModApiStorage::l_get_mod_storage(lua_State *L)
...
-		if (gamedef->registerModStorage(store)) {
-			StorageRef::create(L, store);
-			int object = lua_gettop(L);
-			lua_pushvalue(L, object);
-			return 1;
-		}
-	} else {
+	if (IGameDef *gamedef = getGameDef(L))
+		StorageRef::create(L, new ModMetadata(mod_name, gamedef->getModStorageDatabase()));
+	else {
```
Comment (sfan5) [inline src/script/lua_api/l_storage.cpp:33]:
```suggestion
	if (IGameDef *gamedef = getGameDef(L)) {
		StorageRef::create(L, new ModMetadata(mod_name, gamedef->getModStorageDatabase()));
	} else {
```

**Review by sfan5 — COMMENTED** [review]
(code review only)

File: src/database/database-dummy.cpp (hunk 4 above)
Comment (TurkeyMcMac) [reply src/database/database-dummy.cpp:144]: Yes, the method returns false when the implementation is sure nothing changed, just like `removeModEntry`.

File: src/metadata.cpp (hunk 5 above)
Comment (TurkeyMcMac) [reply src/metadata.cpp:64]: Yes, I believe so. Analyzing the implementation of `resolveString`, it seems that the referenced string in the first parameter is never accessed after the string in the second parameter is modified (if it is modified at all).

File: src/content/mods.cpp (hunk 2 above)
Comment (TurkeyMcMac) [reply src/content/mods.cpp:250]: This method has to be able to return `nullptr` to indicate that no string was found.

File: doc/lua_api.txt (hunk 1 above)
Comment (TurkeyMcMac) [reply doc/lua_api.txt:4873]: It means that, when using a non-deprecated backend, the size of mod storage is not limited by the available RAM. This doesn't include the `dummy` backend, which explicitly keeps everything in RAM. Perhaps I should change this comment.

Comment (TurkeyMcMac) [reply doc/lua_api.txt:4874]: The documentation has been changed.

Comment (sfan5) [reply doc/lua_api.txt:4874]: No I understand what it means, but what are mods expected to do with this information?

Comment (TurkeyMcMac) [reply doc/lua_api.txt:4874]: If they detect this feature, they can basically store as much data as they want in mod storage.

Comment (TurkeyMcMac) [reply doc/lua_api.txt:4874]: They wouldn't have to insecurely require lsqlite unless they needed a full-on relational database.

Comment (sfan5) [reply doc/lua_api.txt:4874]: The problem I see here is that the feature entry only says "_maybe_ it won't all be in RAM". If a mod relies on this and stores lots of data then -- depending on circumstances it cannot determine -- the engine might still run out of RAM.
For a mod the only safe way is to not store much data regardless of this feature flag. Therefore it's not useful.

Comment (TurkeyMcMac) [reply doc/lua_api.txt:4874]: The updated documentation should be more clear about this.

**Review by sfan5 — APPROVED** [review]

File: doc/lua_api.txt (hunk 1 above)
Comment (sfan5) [reply doc/lua_api.txt:4874]: grafik
Maybe this diagram makes it clearer.
The problem I see is that mods cannot know _for sure_ (\*) that storing lots of data is safe.
(\*) It is certainly possible to open world.mt and read the current backend. Aside from being obviously unportable we shouldn't encourage or support this way (API guarantees).

This is fundamentally the same reason why #12167 was rejected.

Comment (TurkeyMcMac) [reply doc/lua_api.txt:4874]: It's unfortunate, but mods can specify in their READMEs that a backend should be used other than "files" or "dummy".

**Comment by TurkeyMcMac** [comment]
I rebased to remove `registerModStorage` and `unregisterModStorage` from `DummyGameDef` (and to fix conflicts.)

**Pull request merged**